src/app/reporting/ReportSchedulerImpl.cpp

```diff
@@ -54,80 +67,39 @@ void ReportSchedulerImpl::OnBecameReportable(ReadHandler * aReadHandler)
     VerifyOrReturn(nullptr != node);
 
     Milliseconds32 newTimeout;
-    if (node->IsReportableNow())
-    {
-        // If the handler is reportable now, just schedule a report immediately
-        newTimeout = Milliseconds32(0);
-    }
-    else
-    {
-        // If the handler is not reportable now, schedule a report for the min interval
-        newTimeout = node->GetMinTimestamp() - mTimerDelegate->GetCurrentMonotonicTimestamp();
-    }
-
+    CalculateNextReportTimeout(newTimeout, node);
     ScheduleReport(newTimeout, node);
 }
 
 void ReportSchedulerImpl::OnSubscriptionAction(ReadHandler * apReadHandler)
 {
     ReadHandlerNode * node = FindReadHandlerNode(apReadHandler);
     VerifyOrReturn(nullptr != node);
-    // Schedule callback for max interval by computing the difference between the max timestamp and the current timestamp
     node->SetIntervalTimeStamps(apReadHandler);
-    Milliseconds32 newTimeout = node->GetMaxTimestamp() - mTimerDelegate->GetCurrentMonotonicTimestamp();
+    Milliseconds32 newTimeout;
+    CalculateNextReportTimeout(newTimeout, node);
     ScheduleReport(newTimeout, node);
+    node->SetEngineRunScheduled(false);
```

Comment:
But we don't do this on OnBecameReportable?

I have to admit to being pretty confused about what this "run scheduled" boolean represents and have very low confidence that it works "right"....

Reply:
OnSubscriptionAction being called means an engine run was completed for the current handler, so it can start contributing to calculation the next timeout from that point on.

We should discuss potential (probable) bugfix here if that logic doesn't seem to work: https://github.com/project-chip/connectedhomeip/issues/28052